src/ops/mod.rs

```diff
@@ -1100,24 +1101,48 @@ pub fn get_index_url(crates_file: &Path, registry: &str) -> Result<(String, Cow<
         cur_source = "crates-io".into();
     }
 
-    if let Some(source) = config.get("source") {
-        for (name, v) in source.as_table().ok_or(Cow::Borrowed("source not table"))? {
-            if let Some(replacement) = v.get("replace-with") {
-                replacements.insert(&name[..],
-                                    replacement.as_str().ok_or_else(|| format!("source.{}.replacement not string", name))?);
+    if let Some(toml::Value::Table(source)) = config.get("source") {
+        for (name, v) in source {
+            if let Some(toml::Value::String(replacement)) = v.get("replace-with") {
+                replacements.insert(&name[..], replacement);
             }
 
-            if let Some(url) = v.get("registry") {
-                let url = url.as_str().ok_or_else(|| format!("source.{}.registry not string", name))?.to_string().into();
-                if cur_source == url {
+            if let Some(toml::Value::String(url)) = v.get("registry") {
+                if cur_source == **url {
                     cur_source = name.into();
                 }
 
-                registries.insert(&name[..], url);
+                registries.insert(name, url.into());
             }
         }
     }
 
+    if let Some(toml::Value::Table(table)) = config.get("registries") {
+        table
+            .iter()
+            .filter(|(_, val)| match val {
+				toml::Value::Table(_) => true,
+				_ => false,
+			})
+            .map(|(name, val)| (name, val.as_table().and_then(|table| table.get("index"))))
+            .map(|(name, index)| {
+                (
+                    name,
+                    match index {
+                        Some(toml::Value::String(s)) => Some(s),
+                        _ => None,
+                    },
+                )
+            })
+            .filter(|(_, url)| url.is_some())
+            .map(|(name, url)| (name, url.unwrap()))
+            .for_each(|(name, url)| {
+                if cur_source == **url {
+                    cur_source = name.into()
+                }
+                registries.insert(name, url.into());
+            })
```

Comment:
what a mess, how is this different from
```rust
        for (name, url) in table.iter()
            .flat_map(|(name, val)| val.as_table().and_then(|t| t.get("index")).and_then(|idx| idx.as_str()).map(|v| (name, v))) {
            if cur_source == url {
                cur_source = name.into()
            }
            registries.insert(name, url.into());
        }
```
or, indeed,
```rust
        for (name, url) in table.iter().flat_map(|(name, val)| val.as_table()?.get("index")?.as_str().map(|v| (name, v))) {
            if cur_source == url {
                cur_source = name.into()
            }
            registries.insert(name, url.into());
        }
```
?

Reply:
Your solution is pretty neat. I've updated this PR.